bundles/org.eclipse.osgi/container/src/org/eclipse/osgi/internal/hooks/ServiceLoaderMediatorHook.java

```diff
@@ -0,0 +1,169 @@
+package org.eclipse.osgi.internal.hooks;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.lang.reflect.Method;
+import java.net.URL;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Enumeration;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.ServiceLoader;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+import org.eclipse.osgi.internal.hookregistry.ClassLoaderHook;
+import org.eclipse.osgi.internal.loader.ModuleClassLoader;
+import org.osgi.framework.Constants;
+import org.osgi.framework.wiring.BundleRequirement;
+import org.osgi.framework.wiring.BundleWire;
+import org.osgi.framework.wiring.BundleWiring;
+
+public class ServiceLoaderMediatorHook extends ClassLoaderHook {
+
+	@SuppressWarnings({ "unchecked", "rawtypes" })
+	private Supplier<Optional<Class<?>>> getCallerClassComputer() {
+		try { // The Java-9+ way
```

Comment:
If you place this in a framework fragment then we can make it depend on a higher osgi.ee than Java 8.